{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gumpy, numpy, copy, piezo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('1.25.2', '1.2.4')"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "numpy.__version__, gumpy.__version__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The key object in `gumpy` is the `Genome` which you instantiate with a GenBank file -- this can take 1-2 min"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference = gumpy.Genome('NC_000962.3.gbk')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NC_000962\n",
       "NC_000962.3\n",
       "Mycobacterium tuberculosis H37Rv, complete genome\n",
       "4411532 bases\n",
       "ttgacc...acgtcg\n",
       "metadata for all genes/loci have been included"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reference"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One of the things you can do with a `Genome` object is build `Gene` object(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "pnca_reference = reference.build_gene('pncA')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`__repr__`  is overloaded so you can do this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pncA gene\n",
       "601 nucleotides, codes for protein\n",
       "['g' 'c' 'a' ... 'c' 'g' 't']\n",
       "[-40 -39 -38 ...  -3  -2  -1]\n",
       "['M' 'R' 'A' ... 'S' 'S' '!']\n",
       "[  1   2   3 ... 185 186 187]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pnca_reference"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then various propoerties, like the amino acid sequence is natively exposed as a `numpy` array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['M', 'R', 'A', 'L', 'I', 'I', 'V', 'D', 'V', 'Q', 'N', 'D', 'F',\n",
       "       'C', 'E', 'G', 'G', 'S', 'L', 'A', 'V', 'T', 'G', 'G', 'A', 'A',\n",
       "       'L', 'A', 'R', 'A', 'I', 'S', 'D', 'Y', 'L', 'A', 'E', 'A', 'A',\n",
       "       'D', 'Y', 'H', 'H', 'V', 'V', 'A', 'T', 'K', 'D', 'F', 'H', 'I',\n",
       "       'D', 'P', 'G', 'D', 'H', 'F', 'S', 'G', 'T', 'P', 'D', 'Y', 'S',\n",
       "       'S', 'S', 'W', 'P', 'P', 'H', 'C', 'V', 'S', 'G', 'T', 'P', 'G',\n",
       "       'A', 'D', 'F', 'H', 'P', 'S', 'L', 'D', 'T', 'S', 'A', 'I', 'E',\n",
       "       'A', 'V', 'F', 'Y', 'K', 'G', 'A', 'Y', 'T', 'G', 'A', 'Y', 'S',\n",
       "       'G', 'F', 'E', 'G', 'V', 'D', 'E', 'N', 'G', 'T', 'P', 'L', 'L',\n",
       "       'N', 'W', 'L', 'R', 'Q', 'R', 'G', 'V', 'D', 'E', 'V', 'D', 'V',\n",
       "       'V', 'G', 'I', 'A', 'T', 'D', 'H', 'C', 'V', 'R', 'Q', 'T', 'A',\n",
       "       'E', 'D', 'A', 'V', 'R', 'N', 'G', 'L', 'A', 'T', 'R', 'V', 'L',\n",
       "       'V', 'D', 'L', 'T', 'A', 'G', 'V', 'S', 'A', 'D', 'T', 'T', 'V',\n",
       "       'A', 'A', 'L', 'E', 'E', 'M', 'R', 'T', 'A', 'S', 'V', 'E', 'L',\n",
       "       'V', 'C', 'S', 'S', '!'], dtype='<U1')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pnca_reference.amino_acid_sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Or we can make a simple string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVATKDFHIDPGDHFSGTPDYSSSWPPHCVSGTPGADFHPSLDTSAIEAVFYKGAYTGAYSGFEGVDENGTPLLNWLRQRGVDEVDVVGIATDHCVRQTAEDAVRNGLATRVLVDLTAGVSADTTVAALEEMRTASVELVCSS!'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "''.join(i for i in pnca_reference.amino_acid_sequence)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Because the gene has associated with it a promoter region, we have to remove that to get at the nucleotide sequence of the CDS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "atgcgggcgttgatcatcgtcgacgtgcagaacgacttctgcgagggtggctcgctggcggtaaccggtggcgccgcgctggcccgcgccatcagcgactacctggccgaagcggcggactaccatcacgtcgtggcaaccaaggacttccacatcgacccgggtgaccacttctccggcacaccggactattcctcgtcgtggccaccgcattgcgtcagcggtactcccggcgcggacttccatcccagtctggacacgtcggcaatcgaggcggtgttctacaagggtgcctacaccggagcgtacagcggcttcgaaggagtcgacgagaacggcacgccactgctgaattggctgcggcaacgcggcgtcgatgaggtcgatgtggtcggtattgccaccgatcattgtgtgcgccagacggccgaggacgcggtacgcaatggcttggccaccagggtgctggtggacctgacagcgggtgtgtcggccgataccaccgtcgccgcgctggaggagatgcgcaccgccagcgtcgagttggtttgcagctcctga\n"
     ]
    }
   ],
   "source": [
    "pnca_sequence = ''.join(i for i in pnca_reference.nucleotide_sequence[pnca_reference.nucleotide_number>0])\n",
    "print(pnca_sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we want to work out all 1SNP possible mutations we need to know the codon table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "aminoacids = 'FFLLSSSSYY!!CC!WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG'\n",
    "bases = ['t', 'c', 'a', 'g']\n",
    "all_codons = numpy.array([a+b+c for a in bases for b in bases for c in bases])\n",
    "codon_to_amino_acid = dict(zip(all_codons, aminoacids))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that the below won't allow a stop codon (here `!`) to be introduced"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "index = 1\n",
    "\n",
    "maximum_amino_acids = []\n",
    "amino_acid_position = []\n",
    "possible_mutations = {}\n",
    "\n",
    "for codon in [pnca_sequence[i:i+3] for i in range(0,len(pnca_sequence),3)]:\n",
    "    \n",
    "    aminoacid = codon_to_amino_acid[codon]\n",
    "    mutations = []\n",
    "\n",
    "    for pos in [0, 1, 2]:\n",
    "        for i in bases:\n",
    "            if codon[pos] != i:\n",
    "                new_codon = codon[:pos] + i + codon[pos+1:]\n",
    "                new_aminoacid = codon_to_amino_acid[new_codon]\n",
    "\n",
    "                if new_aminoacid != aminoacid and new_aminoacid not in mutations and new_aminoacid!=\"!\":\n",
    "                    mutations.append(new_aminoacid)\n",
    "\n",
    "    possible_mutations[index] = mutations\n",
    "    maximum_amino_acids.append(len(mutations))\n",
    "    amino_acid_position.append(index)\n",
    "    \n",
    "    index+=1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This gives us a `dict` of all the possible amino acid mutations that can arise at each position through a single nucleotide change"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1: ['L', 'V', 'T', 'K', 'R', 'I'],\n",
       " 2: ['W', 'G', 'L', 'P', 'Q'],\n",
       " 3: ['S', 'P', 'T', 'V', 'E', 'G'],\n",
       " 4: ['M', 'V', 'S', 'W', 'F'],\n",
       " 5: ['F', 'L', 'V', 'T', 'N', 'S', 'M'],\n",
       " 6: ['F', 'L', 'V', 'T', 'N', 'S', 'M'],\n",
       " 7: ['F', 'L', 'I', 'A', 'D', 'G'],\n",
       " 8: ['Y', 'H', 'N', 'V', 'A', 'G', 'E'],\n",
       " 9: ['L', 'M', 'A', 'E', 'G'],\n",
       " 10: ['K', 'E', 'L', 'P', 'R', 'H'],\n",
       " 11: ['Y', 'H', 'D', 'I', 'T', 'S', 'K'],\n",
       " 12: ['Y', 'H', 'N', 'V', 'A', 'G', 'E'],\n",
       " 13: ['L', 'I', 'V', 'S', 'Y', 'C'],\n",
       " 14: ['R', 'S', 'G', 'F', 'Y', 'W'],\n",
       " 15: ['Q', 'K', 'V', 'A', 'G', 'D'],\n",
       " 16: ['C', 'R', 'S', 'V', 'A', 'D'],\n",
       " 17: ['C', 'R', 'S', 'V', 'A', 'D'],\n",
       " 18: ['P', 'T', 'A', 'L', 'W'],\n",
       " 19: ['M', 'V', 'P', 'Q', 'R'],\n",
       " 20: ['S', 'P', 'T', 'V', 'E', 'G'],\n",
       " 21: ['L', 'I', 'A', 'E', 'G'],\n",
       " 22: ['S', 'P', 'A', 'I', 'N'],\n",
       " 23: ['C', 'R', 'S', 'V', 'A', 'D'],\n",
       " 24: ['C', 'R', 'S', 'V', 'A', 'D'],\n",
       " 25: ['S', 'P', 'T', 'V', 'D', 'G'],\n",
       " 26: ['S', 'P', 'T', 'V', 'E', 'G'],\n",
       " 27: ['M', 'V', 'P', 'Q', 'R'],\n",
       " 28: ['S', 'P', 'T', 'V', 'D', 'G'],\n",
       " 29: ['C', 'S', 'G', 'L', 'P', 'H'],\n",
       " 30: ['S', 'P', 'T', 'V', 'D', 'G'],\n",
       " 31: ['F', 'L', 'V', 'T', 'N', 'S', 'M'],\n",
       " 32: ['C', 'R', 'G', 'I', 'T', 'N'],\n",
       " 33: ['Y', 'H', 'N', 'V', 'A', 'G', 'E'],\n",
       " 34: ['H', 'N', 'D', 'F', 'S', 'C'],\n",
       " 35: ['M', 'V', 'P', 'Q', 'R'],\n",
       " 36: ['S', 'P', 'T', 'V', 'D', 'G'],\n",
       " 37: ['Q', 'K', 'V', 'A', 'G', 'D'],\n",
       " 38: ['S', 'P', 'T', 'V', 'E', 'G'],\n",
       " 39: ['S', 'P', 'T', 'V', 'E', 'G'],\n",
       " 40: ['Y', 'H', 'N', 'V', 'A', 'G', 'E'],\n",
       " 41: ['H', 'N', 'D', 'F', 'S', 'C'],\n",
       " 42: ['Y', 'N', 'D', 'L', 'P', 'R', 'Q'],\n",
       " 43: ['Y', 'N', 'D', 'L', 'P', 'R', 'Q'],\n",
       " 44: ['F', 'L', 'I', 'A', 'D', 'G'],\n",
       " 45: ['L', 'M', 'A', 'E', 'G'],\n",
       " 46: ['S', 'P', 'T', 'V', 'E', 'G'],\n",
       " 47: ['S', 'P', 'A', 'I', 'N'],\n",
       " 48: ['Q', 'E', 'M', 'T', 'R', 'N'],\n",
       " 49: ['Y', 'H', 'N', 'V', 'A', 'G', 'E'],\n",
       " 50: ['L', 'I', 'V', 'S', 'Y', 'C'],\n",
       " 51: ['Y', 'N', 'D', 'L', 'P', 'R', 'Q'],\n",
       " 52: ['F', 'L', 'V', 'T', 'N', 'S', 'M'],\n",
       " 53: ['Y', 'H', 'N', 'V', 'A', 'G', 'E'],\n",
       " 54: ['S', 'T', 'A', 'L', 'Q', 'R'],\n",
       " 55: ['C', 'R', 'S', 'V', 'A', 'D'],\n",
       " 56: ['Y', 'H', 'N', 'V', 'A', 'G', 'E'],\n",
       " 57: ['Y', 'N', 'D', 'L', 'P', 'R', 'Q'],\n",
       " 58: ['L', 'I', 'V', 'S', 'Y', 'C'],\n",
       " 59: ['P', 'T', 'A', 'F', 'Y', 'C'],\n",
       " 60: ['C', 'R', 'S', 'V', 'A', 'D'],\n",
       " 61: ['S', 'P', 'A', 'I', 'K', 'R'],\n",
       " 62: ['S', 'T', 'A', 'L', 'Q', 'R'],\n",
       " 63: ['Y', 'H', 'N', 'V', 'A', 'G', 'E'],\n",
       " 64: ['H', 'N', 'D', 'F', 'S', 'C'],\n",
       " 65: ['P', 'T', 'A', 'F', 'Y', 'C'],\n",
       " 66: ['P', 'T', 'A', 'L', 'W'],\n",
       " 67: ['P', 'T', 'A', 'L', 'W'],\n",
       " 68: ['R', 'G', 'L', 'S', 'C'],\n",
       " 69: ['S', 'T', 'A', 'L', 'Q', 'R'],\n",
       " 70: ['S', 'T', 'A', 'L', 'Q', 'R'],\n",
       " 71: ['Y', 'N', 'D', 'L', 'P', 'R', 'Q'],\n",
       " 72: ['R', 'S', 'G', 'F', 'Y', 'W'],\n",
       " 73: ['F', 'L', 'I', 'A', 'D', 'G'],\n",
       " 74: ['C', 'R', 'G', 'I', 'T', 'N'],\n",
       " 75: ['C', 'R', 'S', 'V', 'A', 'D'],\n",
       " 76: ['S', 'P', 'A', 'I', 'N'],\n",
       " 77: ['S', 'T', 'A', 'L', 'H', 'R'],\n",
       " 78: ['C', 'R', 'S', 'V', 'A', 'D'],\n",
       " 79: ['S', 'P', 'T', 'V', 'E', 'G'],\n",
       " 80: ['Y', 'H', 'N', 'V', 'A', 'G', 'E'],\n",
       " 81: ['L', 'I', 'V', 'S', 'Y', 'C'],\n",
       " 82: ['Y', 'N', 'D', 'L', 'P', 'R', 'Q'],\n",
       " 83: ['S', 'T', 'A', 'L', 'H', 'R'],\n",
       " 84: ['C', 'R', 'G', 'I', 'T', 'N'],\n",
       " 85: ['M', 'V', 'P', 'Q', 'R'],\n",
       " 86: ['Y', 'H', 'N', 'V', 'A', 'G', 'E'],\n",
       " 87: ['S', 'P', 'A', 'M', 'K', 'R'],\n",
       " 88: ['P', 'T', 'A', 'L', 'W'],\n",
       " 89: ['S', 'P', 'T', 'V', 'E', 'G'],\n",
       " 90: ['F', 'L', 'V', 'T', 'N', 'S', 'M'],\n",
       " 91: ['Q', 'K', 'V', 'A', 'G', 'D'],\n",
       " 92: ['S', 'P', 'T', 'V', 'E', 'G'],\n",
       " 93: ['L', 'M', 'A', 'E', 'G'],\n",
       " 94: ['L', 'I', 'V', 'S', 'Y', 'C'],\n",
       " 95: ['H', 'N', 'D', 'F', 'S', 'C'],\n",
       " 96: ['Q', 'E', 'M', 'T', 'R', 'N'],\n",
       " 97: ['C', 'R', 'S', 'V', 'A', 'D'],\n",
       " 98: ['S', 'P', 'T', 'V', 'D', 'G'],\n",
       " 99: ['H', 'N', 'D', 'F', 'S', 'C'],\n",
       " 100: ['S', 'P', 'A', 'I', 'N'],\n",
       " 101: ['R', 'V', 'A', 'E'],\n",
       " 102: ['S', 'P', 'T', 'V', 'E', 'G'],\n",
       " 103: ['H', 'N', 'D', 'F', 'S', 'C'],\n",
       " 104: ['C', 'R', 'G', 'I', 'T', 'N'],\n",
       " 105: ['C', 'R', 'S', 'V', 'A', 'D'],\n",
       " 106: ['L', 'I', 'V', 'S', 'Y', 'C'],\n",
       " 107: ['Q', 'K', 'V', 'A', 'G', 'D'],\n",
       " 108: ['R', 'V', 'A', 'E'],\n",
       " 109: ['F', 'L', 'I', 'A', 'D', 'G'],\n",
       " 110: ['Y', 'H', 'N', 'V', 'A', 'G', 'E'],\n",
       " 111: ['Q', 'K', 'V', 'A', 'G', 'D'],\n",
       " 112: ['Y', 'H', 'D', 'I', 'T', 'S', 'K'],\n",
       " 113: ['C', 'R', 'S', 'V', 'A', 'D'],\n",
       " 114: ['S', 'P', 'A', 'M', 'K', 'R'],\n",
       " 115: ['S', 'T', 'A', 'L', 'Q', 'R'],\n",
       " 116: ['M', 'V', 'P', 'Q', 'R'],\n",
       " 117: ['M', 'V', 'P', 'Q', 'R'],\n",
       " 118: ['Y', 'H', 'D', 'I', 'T', 'S', 'K'],\n",
       " 119: ['R', 'G', 'L', 'S', 'C'],\n",
       " 120: ['M', 'V', 'P', 'Q', 'R'],\n",
       " 121: ['W', 'G', 'L', 'P', 'Q'],\n",
       " 122: ['K', 'E', 'L', 'P', 'R', 'H'],\n",
       " 123: ['C', 'S', 'G', 'L', 'P', 'H'],\n",
       " 124: ['C', 'R', 'S', 'V', 'A', 'D'],\n",
       " 125: ['F', 'L', 'I', 'A', 'D', 'G'],\n",
       " 126: ['Y', 'H', 'N', 'V', 'A', 'G', 'E'],\n",
       " 127: ['Q', 'K', 'V', 'A', 'G', 'D'],\n",
       " 128: ['F', 'L', 'I', 'A', 'D', 'G'],\n",
       " 129: ['Y', 'H', 'N', 'V', 'A', 'G', 'E'],\n",
       " 130: ['L', 'M', 'A', 'E', 'G'],\n",
       " 131: ['F', 'L', 'I', 'A', 'D', 'G'],\n",
       " 132: ['C', 'R', 'S', 'V', 'A', 'D'],\n",
       " 133: ['F', 'L', 'V', 'T', 'N', 'S', 'M'],\n",
       " 134: ['S', 'P', 'T', 'V', 'D', 'G'],\n",
       " 135: ['S', 'P', 'A', 'I', 'N'],\n",
       " 136: ['Y', 'H', 'N', 'V', 'A', 'G', 'E'],\n",
       " 137: ['Y', 'N', 'D', 'L', 'P', 'R', 'Q'],\n",
       " 138: ['R', 'S', 'G', 'F', 'Y', 'W'],\n",
       " 139: ['L', 'M', 'A', 'E', 'G'],\n",
       " 140: ['C', 'S', 'G', 'L', 'P', 'H'],\n",
       " 141: ['K', 'E', 'L', 'P', 'R', 'H'],\n",
       " 142: ['S', 'P', 'A', 'M', 'K', 'R'],\n",
       " 143: ['S', 'P', 'T', 'V', 'D', 'G'],\n",
       " 144: ['Q', 'K', 'V', 'A', 'G', 'D'],\n",
       " 145: ['Y', 'H', 'N', 'V', 'A', 'G', 'E'],\n",
       " 146: ['S', 'P', 'T', 'V', 'E', 'G'],\n",
       " 147: ['L', 'I', 'A', 'E', 'G'],\n",
       " 148: ['C', 'S', 'G', 'L', 'P', 'H'],\n",
       " 149: ['Y', 'H', 'D', 'I', 'T', 'S', 'K'],\n",
       " 150: ['C', 'R', 'S', 'V', 'A', 'D'],\n",
       " 151: ['M', 'V', 'S', 'W', 'F'],\n",
       " 152: ['S', 'P', 'T', 'V', 'D', 'G'],\n",
       " 153: ['S', 'P', 'A', 'I', 'N'],\n",
       " 154: ['W', 'G', 'M', 'T', 'K', 'S'],\n",
       " 155: ['L', 'M', 'A', 'E', 'G'],\n",
       " 156: ['M', 'V', 'P', 'Q', 'R'],\n",
       " 157: ['L', 'M', 'A', 'E', 'G'],\n",
       " 158: ['Y', 'H', 'N', 'V', 'A', 'G', 'E'],\n",
       " 159: ['M', 'V', 'P', 'Q', 'R'],\n",
       " 160: ['S', 'P', 'A', 'I', 'K', 'R'],\n",
       " 161: ['S', 'P', 'T', 'V', 'E', 'G'],\n",
       " 162: ['C', 'R', 'S', 'V', 'A', 'D'],\n",
       " 163: ['L', 'M', 'A', 'E', 'G'],\n",
       " 164: ['P', 'T', 'A', 'L', 'W'],\n",
       " 165: ['S', 'P', 'T', 'V', 'D', 'G'],\n",
       " 166: ['Y', 'H', 'N', 'V', 'A', 'G', 'E'],\n",
       " 167: ['S', 'P', 'A', 'I', 'N'],\n",
       " 168: ['S', 'P', 'A', 'I', 'N'],\n",
       " 169: ['F', 'L', 'I', 'A', 'D', 'G'],\n",
       " 170: ['S', 'P', 'T', 'V', 'D', 'G'],\n",
       " 171: ['S', 'P', 'T', 'V', 'E', 'G'],\n",
       " 172: ['M', 'V', 'P', 'Q', 'R'],\n",
       " 173: ['Q', 'K', 'V', 'A', 'G', 'D'],\n",
       " 174: ['Q', 'K', 'V', 'A', 'G', 'D'],\n",
       " 175: ['L', 'V', 'T', 'K', 'R', 'I'],\n",
       " 176: ['C', 'S', 'G', 'L', 'P', 'H'],\n",
       " 177: ['S', 'P', 'A', 'I', 'N'],\n",
       " 178: ['S', 'P', 'T', 'V', 'D', 'G'],\n",
       " 179: ['C', 'R', 'G', 'I', 'T', 'N'],\n",
       " 180: ['F', 'L', 'I', 'A', 'D', 'G'],\n",
       " 181: ['Q', 'K', 'V', 'A', 'G', 'D'],\n",
       " 182: ['M', 'V', 'S', 'W', 'F'],\n",
       " 183: ['F', 'L', 'I', 'A', 'D', 'G'],\n",
       " 184: ['R', 'S', 'G', 'F', 'Y', 'W'],\n",
       " 185: ['C', 'R', 'G', 'I', 'T', 'N'],\n",
       " 186: ['P', 'T', 'A', 'F', 'Y', 'C'],\n",
       " 187: ['R', 'G', 'L', 'S', 'C', 'W']}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "possible_mutations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We don't have to use `piezo` to load the resistance catalogue as it is just a CSV file but it does create some extra columns for us which makes it easier to select the rows we want"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>PREDICTION</th>\n",
       "      <th>SOURCE</th>\n",
       "      <th>EVIDENCE</th>\n",
       "      <th>OTHER</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>GENE</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>MUTATION_AFFECTS</th>\n",
       "      <th>MUTATION_TYPE</th>\n",
       "      <th>MINOR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RIF</td>\n",
       "      <td>1274_del_cggcac</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'Present_SOLO_R': 0, 'Present_SOLO_SR': 1, 'P...</td>\n",
       "      <td>{'FINAL_CONFIDENCE_GRADING': '2) Assoc w R - I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1274</td>\n",
       "      <td>CDS</td>\n",
       "      <td>INDEL</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RIF</td>\n",
       "      <td>1277_del_caccagcca</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'Present_SOLO_R': 1, 'Present_SOLO_SR': 1, 'P...</td>\n",
       "      <td>{'FINAL_CONFIDENCE_GRADING': '2) Assoc w R - I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1277</td>\n",
       "      <td>CDS</td>\n",
       "      <td>INDEL</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RIF</td>\n",
       "      <td>1278_del_accagc</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'Present_SOLO_R': 1, 'Present_SOLO_SR': 1, 'P...</td>\n",
       "      <td>{'FINAL_CONFIDENCE_GRADING': '2) Assoc w R - I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1278</td>\n",
       "      <td>CDS</td>\n",
       "      <td>INDEL</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RIF</td>\n",
       "      <td>1278_del_accagccagctg</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'Present_SOLO_R': 0, 'Present_SOLO_SR': 0, 'P...</td>\n",
       "      <td>{'FINAL_CONFIDENCE_GRADING': '2) Assoc w R - I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1278</td>\n",
       "      <td>CDS</td>\n",
       "      <td>INDEL</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RIF</td>\n",
       "      <td>1279_del_ccagcc</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'Present_SOLO_R': 0, 'Present_SOLO_SR': 0, 'P...</td>\n",
       "      <td>{'FINAL_CONFIDENCE_GRADING': '2) Assoc w R - I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1279</td>\n",
       "      <td>CDS</td>\n",
       "      <td>INDEL</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1685</th>\n",
       "      <td>STM</td>\n",
       "      <td>del_0.0</td>\n",
       "      <td>U</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>None</td>\n",
       "      <td>GENE</td>\n",
       "      <td>INDEL</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1686</th>\n",
       "      <td>ETH</td>\n",
       "      <td>del_0.0</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ethA</td>\n",
       "      <td>None</td>\n",
       "      <td>GENE</td>\n",
       "      <td>INDEL</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1687</th>\n",
       "      <td>CAP</td>\n",
       "      <td>del_0.0</td>\n",
       "      <td>U</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tlyA</td>\n",
       "      <td>None</td>\n",
       "      <td>GENE</td>\n",
       "      <td>INDEL</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1688</th>\n",
       "      <td>ETH</td>\n",
       "      <td>del_0.0</td>\n",
       "      <td>U</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>fabG1</td>\n",
       "      <td>None</td>\n",
       "      <td>GENE</td>\n",
       "      <td>INDEL</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1689</th>\n",
       "      <td>INH</td>\n",
       "      <td>del_0.0</td>\n",
       "      <td>U</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>fabG1</td>\n",
       "      <td>None</td>\n",
       "      <td>GENE</td>\n",
       "      <td>INDEL</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1690 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     DRUG               MUTATION PREDICTION SOURCE  \\\n",
       "0     RIF        1274_del_cggcac          R     {}   \n",
       "1     RIF     1277_del_caccagcca          R     {}   \n",
       "2     RIF        1278_del_accagc          R     {}   \n",
       "3     RIF  1278_del_accagccagctg          R     {}   \n",
       "4     RIF        1279_del_ccagcc          R     {}   \n",
       "...   ...                    ...        ...    ...   \n",
       "1685  STM                del_0.0          U     {}   \n",
       "1686  ETH                del_0.0          R     {}   \n",
       "1687  CAP                del_0.0          U     {}   \n",
       "1688  ETH                del_0.0          U     {}   \n",
       "1689  INH                del_0.0          U     {}   \n",
       "\n",
       "                                               EVIDENCE  \\\n",
       "0     {'Present_SOLO_R': 0, 'Present_SOLO_SR': 1, 'P...   \n",
       "1     {'Present_SOLO_R': 1, 'Present_SOLO_SR': 1, 'P...   \n",
       "2     {'Present_SOLO_R': 1, 'Present_SOLO_SR': 1, 'P...   \n",
       "3     {'Present_SOLO_R': 0, 'Present_SOLO_SR': 0, 'P...   \n",
       "4     {'Present_SOLO_R': 0, 'Present_SOLO_SR': 0, 'P...   \n",
       "...                                                 ...   \n",
       "1685                                                 {}   \n",
       "1686                                                 {}   \n",
       "1687                                                 {}   \n",
       "1688                                                 {}   \n",
       "1689                                                 {}   \n",
       "\n",
       "                                                  OTHER  Unnamed: 0   GENE  \\\n",
       "0     {'FINAL_CONFIDENCE_GRADING': '2) Assoc w R - I...         NaN   rpoB   \n",
       "1     {'FINAL_CONFIDENCE_GRADING': '2) Assoc w R - I...         NaN   rpoB   \n",
       "2     {'FINAL_CONFIDENCE_GRADING': '2) Assoc w R - I...         NaN   rpoB   \n",
       "3     {'FINAL_CONFIDENCE_GRADING': '2) Assoc w R - I...         NaN   rpoB   \n",
       "4     {'FINAL_CONFIDENCE_GRADING': '2) Assoc w R - I...         NaN   rpoB   \n",
       "...                                                 ...         ...    ...   \n",
       "1685                                                 {}         NaN   rpsL   \n",
       "1686                                                 {}         NaN   ethA   \n",
       "1687                                                 {}         NaN   tlyA   \n",
       "1688                                                 {}         NaN  fabG1   \n",
       "1689                                                 {}         NaN  fabG1   \n",
       "\n",
       "     POSITION MUTATION_AFFECTS MUTATION_TYPE MINOR  \n",
       "0        1274              CDS         INDEL        \n",
       "1        1277              CDS         INDEL        \n",
       "2        1278              CDS         INDEL        \n",
       "3        1278              CDS         INDEL        \n",
       "4        1279              CDS         INDEL        \n",
       "...       ...              ...           ...   ...  \n",
       "1685     None             GENE         INDEL        \n",
       "1686     None             GENE         INDEL        \n",
       "1687     None             GENE         INDEL        \n",
       "1688     None             GENE         INDEL        \n",
       "1689     None             GENE         INDEL        \n",
       "\n",
       "[1690 rows x 12 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "catalogue = piezo.ResistanceCatalogue('NC_000962.3_WHO-UCN-GTB-PCI-2021.7_v1.0_GARC1_RUS.csv')\n",
    "catalogue.catalogue.rules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "But the catalogue includes a wide range of rules, including variants known to associated with `S` as well as wild card rules etc.\n",
    "\n",
    "To start with lets keep it simple and focus down to just SNPs in CDS with confer resistance to a drug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "specifc_resistance_mutations = catalogue.catalogue.rules[ (catalogue.catalogue.rules.PREDICTION=='R') &\\\n",
    "                                                          (catalogue.catalogue.rules.MUTATION_TYPE=='SNP') &\\\n",
    "                                                          (catalogue.catalogue.rules.MUTATION_AFFECTS=='CDS') &\\\n",
    "                                                          (catalogue.catalogue.rules.POSITION!='*')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>PREDICTION</th>\n",
       "      <th>SOURCE</th>\n",
       "      <th>EVIDENCE</th>\n",
       "      <th>OTHER</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>GENE</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>MUTATION_AFFECTS</th>\n",
       "      <th>MUTATION_TYPE</th>\n",
       "      <th>MINOR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>RIF</td>\n",
       "      <td>A451G</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'Present_SOLO_R': 0, 'Present_SOLO_SR': 0, 'P...</td>\n",
       "      <td>{'FINAL_CONFIDENCE_GRADING': '2) Assoc w R - I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>451</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>RIF</td>\n",
       "      <td>A451V</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'Present_SOLO_R': 0, 'Present_SOLO_SR': 4, 'P...</td>\n",
       "      <td>{'FINAL_CONFIDENCE_GRADING': '2) Assoc w R - I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>451</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>RIF</td>\n",
       "      <td>D435A</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'Present_SOLO_R': 0, 'Present_SOLO_SR': 1, 'P...</td>\n",
       "      <td>{'FINAL_CONFIDENCE_GRADING': '2) Assoc w R - I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>435</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>RIF</td>\n",
       "      <td>D435E</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'Present_SOLO_R': 0, 'Present_SOLO_SR': 0, 'P...</td>\n",
       "      <td>{'FINAL_CONFIDENCE_GRADING': '2) Assoc w R - I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>435</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>RIF</td>\n",
       "      <td>D435F</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'Present_SOLO_R': 35, 'Present_SOLO_SR': 38, ...</td>\n",
       "      <td>{'FINAL_CONFIDENCE_GRADING': '1) Assoc w R'}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>435</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1347</th>\n",
       "      <td>RIF</td>\n",
       "      <td>A451?</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>451</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1348</th>\n",
       "      <td>RIF</td>\n",
       "      <td>L452?</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>452</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1438</th>\n",
       "      <td>PZA</td>\n",
       "      <td>H71D</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pncA</td>\n",
       "      <td>71</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1439</th>\n",
       "      <td>PZA</td>\n",
       "      <td>L116R</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pncA</td>\n",
       "      <td>116</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1440</th>\n",
       "      <td>PZA</td>\n",
       "      <td>T135N</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pncA</td>\n",
       "      <td>135</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>503 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     DRUG MUTATION PREDICTION SOURCE  \\\n",
       "38    RIF    A451G          R     {}   \n",
       "39    RIF    A451V          R     {}   \n",
       "40    RIF    D435A          R     {}   \n",
       "41    RIF    D435E          R     {}   \n",
       "42    RIF    D435F          R     {}   \n",
       "...   ...      ...        ...    ...   \n",
       "1347  RIF    A451?          R     {}   \n",
       "1348  RIF    L452?          R     {}   \n",
       "1438  PZA     H71D          R     {}   \n",
       "1439  PZA    L116R          R     {}   \n",
       "1440  PZA    T135N          R     {}   \n",
       "\n",
       "                                               EVIDENCE  \\\n",
       "38    {'Present_SOLO_R': 0, 'Present_SOLO_SR': 0, 'P...   \n",
       "39    {'Present_SOLO_R': 0, 'Present_SOLO_SR': 4, 'P...   \n",
       "40    {'Present_SOLO_R': 0, 'Present_SOLO_SR': 1, 'P...   \n",
       "41    {'Present_SOLO_R': 0, 'Present_SOLO_SR': 0, 'P...   \n",
       "42    {'Present_SOLO_R': 35, 'Present_SOLO_SR': 38, ...   \n",
       "...                                                 ...   \n",
       "1347                                                 {}   \n",
       "1348                                                 {}   \n",
       "1438                                                 {}   \n",
       "1439                                                 {}   \n",
       "1440                                                 {}   \n",
       "\n",
       "                                                  OTHER  Unnamed: 0  GENE  \\\n",
       "38    {'FINAL_CONFIDENCE_GRADING': '2) Assoc w R - I...         NaN  rpoB   \n",
       "39    {'FINAL_CONFIDENCE_GRADING': '2) Assoc w R - I...         NaN  rpoB   \n",
       "40    {'FINAL_CONFIDENCE_GRADING': '2) Assoc w R - I...         NaN  rpoB   \n",
       "41    {'FINAL_CONFIDENCE_GRADING': '2) Assoc w R - I...         NaN  rpoB   \n",
       "42         {'FINAL_CONFIDENCE_GRADING': '1) Assoc w R'}         NaN  rpoB   \n",
       "...                                                 ...         ...   ...   \n",
       "1347                                                 {}         NaN  rpoB   \n",
       "1348                                                 {}         NaN  rpoB   \n",
       "1438                                                 {}         NaN  pncA   \n",
       "1439                                                 {}         NaN  pncA   \n",
       "1440                                                 {}         NaN  pncA   \n",
       "\n",
       "     POSITION MUTATION_AFFECTS MUTATION_TYPE MINOR  \n",
       "38        451              CDS           SNP        \n",
       "39        451              CDS           SNP        \n",
       "40        435              CDS           SNP        \n",
       "41        435              CDS           SNP        \n",
       "42        435              CDS           SNP        \n",
       "...       ...              ...           ...   ...  \n",
       "1347      451              CDS           SNP        \n",
       "1348      452              CDS           SNP        \n",
       "1438       71              CDS           SNP        \n",
       "1439      116              CDS           SNP        \n",
       "1440      135              CDS           SNP        \n",
       "\n",
       "[503 rows x 12 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "specifc_resistance_mutations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can subset further to only those mutations in *pncA* which confer resistance to PZA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>PREDICTION</th>\n",
       "      <th>SOURCE</th>\n",
       "      <th>EVIDENCE</th>\n",
       "      <th>OTHER</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>GENE</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>MUTATION_AFFECTS</th>\n",
       "      <th>MUTATION_TYPE</th>\n",
       "      <th>MINOR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>458</th>\n",
       "      <td>PZA</td>\n",
       "      <td>A102P</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'Present_SOLO_R': 5, 'Present_SOLO_SR': 6, 'P...</td>\n",
       "      <td>{'FINAL_CONFIDENCE_GRADING': '1) Assoc w R'}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pncA</td>\n",
       "      <td>102</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459</th>\n",
       "      <td>PZA</td>\n",
       "      <td>A134V</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'Present_SOLO_R': 12, 'Present_SOLO_SR': 13, ...</td>\n",
       "      <td>{'FINAL_CONFIDENCE_GRADING': '1) Assoc w R'}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pncA</td>\n",
       "      <td>134</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460</th>\n",
       "      <td>PZA</td>\n",
       "      <td>A143G</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'Present_SOLO_R': 5, 'Present_SOLO_SR': 5, 'P...</td>\n",
       "      <td>{'FINAL_CONFIDENCE_GRADING': '1) Assoc w R'}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pncA</td>\n",
       "      <td>143</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461</th>\n",
       "      <td>PZA</td>\n",
       "      <td>A146T</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'Present_SOLO_R': 7, 'Present_SOLO_SR': 9, 'P...</td>\n",
       "      <td>{'FINAL_CONFIDENCE_GRADING': '1) Assoc w R'}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pncA</td>\n",
       "      <td>146</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td>PZA</td>\n",
       "      <td>A146V</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'Present_SOLO_R': 15, 'Present_SOLO_SR': 15, ...</td>\n",
       "      <td>{'FINAL_CONFIDENCE_GRADING': '1) Assoc w R'}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pncA</td>\n",
       "      <td>146</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>623</th>\n",
       "      <td>PZA</td>\n",
       "      <td>Y64D</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'Present_SOLO_R': 2, 'Present_SOLO_SR': 4, 'P...</td>\n",
       "      <td>{'FINAL_CONFIDENCE_GRADING': '2) Assoc w R - I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pncA</td>\n",
       "      <td>64</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>624</th>\n",
       "      <td>PZA</td>\n",
       "      <td>Y99!</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'Present_SOLO_R': 2, 'Present_SOLO_SR': 2, 'P...</td>\n",
       "      <td>{'FINAL_CONFIDENCE_GRADING': '2) Assoc w R - I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pncA</td>\n",
       "      <td>99</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1438</th>\n",
       "      <td>PZA</td>\n",
       "      <td>H71D</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pncA</td>\n",
       "      <td>71</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1439</th>\n",
       "      <td>PZA</td>\n",
       "      <td>L116R</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pncA</td>\n",
       "      <td>116</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1440</th>\n",
       "      <td>PZA</td>\n",
       "      <td>T135N</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pncA</td>\n",
       "      <td>135</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>170 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     DRUG MUTATION PREDICTION SOURCE  \\\n",
       "458   PZA    A102P          R     {}   \n",
       "459   PZA    A134V          R     {}   \n",
       "460   PZA    A143G          R     {}   \n",
       "461   PZA    A146T          R     {}   \n",
       "462   PZA    A146V          R     {}   \n",
       "...   ...      ...        ...    ...   \n",
       "623   PZA     Y64D          R     {}   \n",
       "624   PZA     Y99!          R     {}   \n",
       "1438  PZA     H71D          R     {}   \n",
       "1439  PZA    L116R          R     {}   \n",
       "1440  PZA    T135N          R     {}   \n",
       "\n",
       "                                               EVIDENCE  \\\n",
       "458   {'Present_SOLO_R': 5, 'Present_SOLO_SR': 6, 'P...   \n",
       "459   {'Present_SOLO_R': 12, 'Present_SOLO_SR': 13, ...   \n",
       "460   {'Present_SOLO_R': 5, 'Present_SOLO_SR': 5, 'P...   \n",
       "461   {'Present_SOLO_R': 7, 'Present_SOLO_SR': 9, 'P...   \n",
       "462   {'Present_SOLO_R': 15, 'Present_SOLO_SR': 15, ...   \n",
       "...                                                 ...   \n",
       "623   {'Present_SOLO_R': 2, 'Present_SOLO_SR': 4, 'P...   \n",
       "624   {'Present_SOLO_R': 2, 'Present_SOLO_SR': 2, 'P...   \n",
       "1438                                                 {}   \n",
       "1439                                                 {}   \n",
       "1440                                                 {}   \n",
       "\n",
       "                                                  OTHER  Unnamed: 0  GENE  \\\n",
       "458        {'FINAL_CONFIDENCE_GRADING': '1) Assoc w R'}         NaN  pncA   \n",
       "459        {'FINAL_CONFIDENCE_GRADING': '1) Assoc w R'}         NaN  pncA   \n",
       "460        {'FINAL_CONFIDENCE_GRADING': '1) Assoc w R'}         NaN  pncA   \n",
       "461        {'FINAL_CONFIDENCE_GRADING': '1) Assoc w R'}         NaN  pncA   \n",
       "462        {'FINAL_CONFIDENCE_GRADING': '1) Assoc w R'}         NaN  pncA   \n",
       "...                                                 ...         ...   ...   \n",
       "623   {'FINAL_CONFIDENCE_GRADING': '2) Assoc w R - I...         NaN  pncA   \n",
       "624   {'FINAL_CONFIDENCE_GRADING': '2) Assoc w R - I...         NaN  pncA   \n",
       "1438                                                 {}         NaN  pncA   \n",
       "1439                                                 {}         NaN  pncA   \n",
       "1440                                                 {}         NaN  pncA   \n",
       "\n",
       "     POSITION MUTATION_AFFECTS MUTATION_TYPE MINOR  \n",
       "458       102              CDS           SNP        \n",
       "459       134              CDS           SNP        \n",
       "460       143              CDS           SNP        \n",
       "461       146              CDS           SNP        \n",
       "462       146              CDS           SNP        \n",
       "...       ...              ...           ...   ...  \n",
       "623        64              CDS           SNP        \n",
       "624        99              CDS           SNP        \n",
       "1438       71              CDS           SNP        \n",
       "1439      116              CDS           SNP        \n",
       "1440      135              CDS           SNP        \n",
       "\n",
       "[170 rows x 12 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pnca_resistant_mutations = specifc_resistance_mutations[(specifc_resistance_mutations.DRUG=='PZA') & (specifc_resistance_mutations.GENE=='pncA')]\n",
    "pnca_resistant_mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['A102P',\n",
       " 'A134V',\n",
       " 'A143G',\n",
       " 'A146T',\n",
       " 'A146V',\n",
       " 'A171E',\n",
       " 'A3E',\n",
       " 'A46E',\n",
       " 'A46V',\n",
       " 'C138R',\n",
       " 'C14!',\n",
       " 'C14R',\n",
       " 'C72R',\n",
       " 'C72Y',\n",
       " 'D12A',\n",
       " 'D12E',\n",
       " 'D12G',\n",
       " 'D12N',\n",
       " 'D49A',\n",
       " 'D49E',\n",
       " 'D49G',\n",
       " 'D49N',\n",
       " 'D63A',\n",
       " 'D63G',\n",
       " 'D8A',\n",
       " 'D8E',\n",
       " 'D8G',\n",
       " 'D8N',\n",
       " 'E15!',\n",
       " 'E181!',\n",
       " 'E91!',\n",
       " 'F106S',\n",
       " 'F13I',\n",
       " 'F13L',\n",
       " 'F58L',\n",
       " 'F81V',\n",
       " 'F94C',\n",
       " 'F94L',\n",
       " 'F94S',\n",
       " 'G105D',\n",
       " 'G105V',\n",
       " 'G108R',\n",
       " 'G132A',\n",
       " 'G132D',\n",
       " 'G132S',\n",
       " 'G162D',\n",
       " 'G17D',\n",
       " 'G24D',\n",
       " 'G97C',\n",
       " 'G97D',\n",
       " 'G97R',\n",
       " 'G97S',\n",
       " 'G97V',\n",
       " 'H51D',\n",
       " 'H51P',\n",
       " 'H51Q',\n",
       " 'H51R',\n",
       " 'H51Y',\n",
       " 'H57D',\n",
       " 'H57R',\n",
       " 'H57Y',\n",
       " 'H71P',\n",
       " 'H71R',\n",
       " 'H71Y',\n",
       " 'H82R',\n",
       " 'I133T',\n",
       " 'I31S',\n",
       " 'I5S',\n",
       " 'I6T',\n",
       " 'I90S',\n",
       " 'I90T',\n",
       " 'K48E',\n",
       " 'K48T',\n",
       " 'K96E',\n",
       " 'K96Q',\n",
       " 'K96R',\n",
       " 'K96T',\n",
       " 'L116P',\n",
       " 'L120P',\n",
       " 'L120Q',\n",
       " 'L120R',\n",
       " 'L151S',\n",
       " 'L159R',\n",
       " 'L172P',\n",
       " 'L172R',\n",
       " 'L182S',\n",
       " 'L182W',\n",
       " 'L19P',\n",
       " 'L27P',\n",
       " 'L4S',\n",
       " 'L4W',\n",
       " 'L85P',\n",
       " 'L85R',\n",
       " 'M175T',\n",
       " 'M175V',\n",
       " 'M1T',\n",
       " 'P54L',\n",
       " 'P54Q',\n",
       " 'P62L',\n",
       " 'P62S',\n",
       " 'P62T',\n",
       " 'P69L',\n",
       " 'Q10!',\n",
       " 'Q10H',\n",
       " 'Q10P',\n",
       " 'Q10R',\n",
       " 'Q122!',\n",
       " 'Q141!',\n",
       " 'Q141P',\n",
       " 'R154G',\n",
       " 'S104I',\n",
       " 'S104R',\n",
       " 'S164!',\n",
       " 'S164P',\n",
       " 'S18!',\n",
       " 'S59P',\n",
       " 'S67P',\n",
       " 'T135P',\n",
       " 'T142A',\n",
       " 'T142K',\n",
       " 'T142M',\n",
       " 'T160P',\n",
       " 'T168P',\n",
       " 'T177P',\n",
       " 'T47A',\n",
       " 'T47P',\n",
       " 'T76I',\n",
       " 'T76P',\n",
       " 'V125F',\n",
       " 'V125G',\n",
       " 'V128G',\n",
       " 'V130G',\n",
       " 'V131F',\n",
       " 'V131G',\n",
       " 'V139A',\n",
       " 'V139G',\n",
       " 'V139L',\n",
       " 'V155G',\n",
       " 'V155M',\n",
       " 'V180A',\n",
       " 'V180F',\n",
       " 'V180G',\n",
       " 'V21G',\n",
       " 'V44G',\n",
       " 'V7A',\n",
       " 'V7G',\n",
       " 'V7L',\n",
       " 'V9G',\n",
       " 'W119!',\n",
       " 'W119C',\n",
       " 'W119G',\n",
       " 'W119R',\n",
       " 'W68!',\n",
       " 'W68C',\n",
       " 'W68G',\n",
       " 'W68R',\n",
       " 'Y103!',\n",
       " 'Y103C',\n",
       " 'Y103D',\n",
       " 'Y103H',\n",
       " 'Y103S',\n",
       " 'Y34!',\n",
       " 'Y34D',\n",
       " 'Y41!',\n",
       " 'Y64!',\n",
       " 'Y64D',\n",
       " 'Y99!',\n",
       " 'H71D',\n",
       " 'L116R',\n",
       " 'T135N']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(pnca_resistant_mutations.MUTATION)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
